stream.hpp.in

```diff
@@ -227,7 +228,75 @@ readString( std::istream& is,
             const std::string::size_type count,
             const char delim );
 
+  /** @{
+   *  @name Linked x,y template streaming functions
+   */
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+  // Linked x,y template streaming functions
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+  /** @brief Read a value from a stream
+   *
+   *  The readValue function will read a value of type Y from a stream and
+   *  transform it into its corresponding X value. Only those x values that are
+   *  referenced in the map are considered as legal x values.
+   *
+   *  Simply put, the readValue function will read a value of type Y from a
+   *  stream and transform it into its corresponding X value. This function can
+   *  for instance be used in the implementation of iostream operators of
+   *  enumerators for which the values are linked to other types such as
+   *  strings, integers or even doubles. This function is defined as a template
+   *  function so that the same code does not have to be written multiple
+   *  times.
+   *
+   *  @todo investigate the use of a template linker class instead of the map
+   *        to extend the x,y linking beyond one on one values (e.g. linked a
+   *        range of y values to a given x value, etc.).
+   *
+   *  @param[in,out] in   the input stream from which to read
+   *  @param[in,out] x    the x value
+   *  @param[in] values   the map linking y to x values
+   *
+   *  @return The input stream. If the function failed to read one of the values
+   *          from the map, the position in the stream will be unchanged and
+   *          std::ios_base::failbit will be set. In this case the value of x
+   *          will also be unchanged.
+   */
+  template <typename X, typename Y> std::istream& readValue(std::istream& in,
+          X& x, const std::map<Y, X>& values) noexcept;
```

Comment:
For immutable/roughly-constant-size maps, given `Y` is hashable, prefer `std::unordered_map` to `std::map` 

`std::unordered_map` is implemented as hash table with linked-list separate chaining. Consequently, lookup operations are O(1) on average.
`std::map` is implemented as a balanced tree (read as red-black tree). Consequently, lookup operations are O( log_2(n) ) on average

If we stick with the `std::map`, we'll need to be mindful that the `Y` type must define a `<` operator.